File: ai_player.py
import math
from game_config import GAME_CONFIG

class AIPlayer:
    """AI-controlled player that uses simple heuristics to challenge Player 1"""
    
    def __init__(self, player2, player1):
        self.player2 = player2
        self.player1 = player1
        self.last_shot_time = 0
        self.shot_cooldown = 1.0 / GAME_CONFIG['fire_rate']  # Minimum time between shots
        
    def update(self, dt, current_time):
        """Update AI behavior based on game state"""
        # Calculate distances and directions
        dx = self.player1.x - self.player2.x
        dy = self.player1.y - self.player2.y
        distance = math.sqrt(dx*dx + dy*dy)
        
        # Movement logic
        if not self.player2.shield_active:
            # Move towards center line while maintaining distance from player1
            target_x = GAME_CONFIG['tiles_width'] / 2
            target_y = self.player1.y  # Mirror player1's vertical position
            
            # Calculate movement direction
            move_x = 0
            move_y = 0
            
            # Horizontal movement (towards center)
            if self.player2.x > target_x:
                move_x = -1
            elif self.player2.x < target_x:
                move_x = 1
                
            # Vertical movement (mirror player1)
            if self.player2.y > target_y:
                move_y = -1
            elif self.player2.y < target_y:
                move_y = 1
                
            # Apply movement
            self.player2.move(move_x * GAME_CONFIG['player_speed'], 
                            move_y * GAME_CONFIG['player_speed'], 
                            dt, current_time, self.player1)
        
        # Shooting logic
        if (not self.player2.shield_active and 
            distance < GAME_CONFIG['tiles_width'] / 2 and 
            current_time - self.last_shot_time > self.shot_cooldown):
            self.player2.shoot(current_time)
            self.last_shot_time = current_time
            
        # Shield logic
        # Activate shield if player1 is shooting and we're close
        if (len(self.player1.projectiles) > 0 and 
            distance < GAME_CONFIG['tiles_width'] / 3):
            self.player2.shield_active = True
        else:
            self.player2.shield_active = False 

File: web/src/ai/AIPlayer.js
import { GAME_CONFIG } from '../config.js';

/**
 * Simple heuristic AI that controls Player 2.
 * Ported from ai_player.py.
 */
export class AIPlayer {
  constructor(player2, player1, walls) {
    this.player2 = player2;
    this.player1 = player1;
    this.walls = walls;
    this.lastShotTime = 0;
    this.shotCooldown = 1.0 / GAME_CONFIG.fire_rate;
  }

  update(dt, currentTime) {
    const p1 = this.player1;
    const p2 = this.player2;

    const dx = p1.x - p2.x;
    const dy = p1.y - p2.y;
    const distance = Math.sqrt(dx * dx + dy * dy);

    // ── Movement ─────────────────────────────────────────────
    if (!p2.shieldActive) {
      const targetX = GAME_CONFIG.tiles_width / 2;
      const targetY = p1.y; // mirror player 1 vertically

      let moveX = 0;
      let moveY = 0;
      if (p2.x > targetX) moveX = -1;
      else if (p2.x < targetX) moveX = 1;
      if (p2.y > targetY) moveY = -1;
      else if (p2.y < targetY) moveY = 1;

      p2.move(
        moveX * GAME_CONFIG.player_speed,
        moveY * GAME_CONFIG.player_speed,
        dt, currentTime, p1, this.walls,
      );
    }

    // ── Shooting ─────────────────────────────────────────────
    if (!p2.shieldActive &&
        distance < GAME_CONFIG.tiles_width / 2 &&
        currentTime - this.lastShotTime > this.shotCooldown) {
      p2.shoot(currentTime);
      this.lastShotTime = currentTime;
    }

    // ── Shield ───────────────────────────────────────────────
    p2.shieldActive = (
      p1.projectiles.length > 0 &&
      distance < GAME_CONFIG.tiles_width / 3
    );
  }
}
